controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -227,105 +207,67 @@ func (w *Workload) UpdateKubeletConfigMap(ctx context.Context, version semver.Ve
-	configMapKey := ctrlclient.ObjectKey{Name: kubeadmConfigKey, Namespace: metav1.NamespaceSystem}
-	kubeadmConfigMap, err := w.getConfigMap(ctx, configMapKey)
-	if err != nil {
-		return err
-	}
-	config := &kubeadmConfig{ConfigMap: kubeadmConfigMap}
-	changed, err := config.UpdateScheduler(scheduler)
-	if err != nil {
-		return err
-	}
-
-	if !changed {
-		return nil
-	}
-
-	if err := w.Client.Update(ctx, config.ConfigMap); err != nil {
-		return errors.Wrap(err, "error updating kubeadm ConfigMap")
-	}
-	return nil
+func (w *Workload) UpdateSchedulerInKubeadmConfigMap(ctx context.Context, scheduler bootstrapv1.ControlPlaneComponent, version semver.Version) error {
+	return w.updateKubeadmConfigMap(ctx, func(k *kubeadmConfig) (bool, error) {
+		return k.UpdateScheduler(scheduler, version)
+	})
 }
 
 // RemoveMachineFromKubeadmConfigMap removes the entry for the machine from the kubeadm configmap.
-func (w *Workload) RemoveMachineFromKubeadmConfigMap(ctx context.Context, machine *clusterv1.Machine) error {
+func (w *Workload) RemoveMachineFromKubeadmConfigMap(ctx context.Context, machine *clusterv1.Machine, version semver.Version) error {
 	if machine == nil || machine.Status.NodeRef == nil {
 		// Nothing to do, no node for Machine
 		return nil
 	}
 
-	return w.RemoveNodeFromKubeadmConfigMap(ctx, machine.Status.NodeRef.Name)
+	return w.RemoveNodeFromKubeadmConfigMap(ctx, machine.Status.NodeRef.Name, version)
 }
 
 // RemoveNodeFromKubeadmConfigMap removes the entry for the node from the kubeadm configmap.
-func (w *Workload) RemoveNodeFromKubeadmConfigMap(ctx context.Context, name string) error {
+func (w *Workload) RemoveNodeFromKubeadmConfigMap(ctx context.Context, name string, version semver.Version) error {
 	return util.Retry(func() (bool, error) {
-		configMapKey := ctrlclient.ObjectKey{Name: kubeadmConfigKey, Namespace: metav1.NamespaceSystem}
-		kubeadmConfigMap, err := w.getConfigMap(ctx, configMapKey)
-		if err != nil {
-			Log.Error(err, "unable to get kubeadmConfigMap")
-			return false, nil
-		}
-		config := &kubeadmConfig{ConfigMap: kubeadmConfigMap}
-		if err := config.RemoveAPIEndpoint(name); err != nil {
+		if err := w.updateKubeadmConfigMap(ctx, func(k *kubeadmConfig) (bool, error) {
+			return k.RemoveAPIEndpoint(name, version)
+		}); err != nil {
 			return false, err
 		}
-		if err := w.Client.Update(ctx, config.ConfigMap); err != nil {
-			Log.Error(err, "error updating kubeadm ConfigMap")
-			return false, nil
-		}
 		return true, nil
 	}, 5)
 }
 
+func (w *Workload) updateKubeadmConfigMap(ctx context.Context, mutator func(*kubeadmConfig) (bool, error)) error {
+	configMapKey := ctrlclient.ObjectKey{Name: kubeadmConfigKey, Namespace: metav1.NamespaceSystem}
+	kubeadmConfigMap, err := w.getConfigMap(ctx, configMapKey)
+	if err != nil {
+		return errors.Wrap(err, "failed to get kubeadmConfigMap")
+	}
+	config := &kubeadmConfig{ConfigMap: kubeadmConfigMap}
+	changed, err := mutator(config)
```

Comment:
Why not detect the changes on this end rather than relying on the `mutator` caller to tell us if the map has changed or not?